modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/near/GridNearTxLocal.java

```diff
@@ -3976,6 +3985,9 @@ private String completedTransactionDump(
 
         mapExplicitLocks();
 
+        if (cctx.kernalContext().deploy().enabled() && deploymentLdrId != null)
+            U.restoreDeploymentContext(cctx.kernalContext(), deploymentLdrId);
```

Comment:
Could you please explain why we need restoreDeploymentContext() here?

Reply:
Because OPTIMISTIC transaction can execute an invoke operation on prepare phase. Due to into the operation we will use a work with p2p classes.
If you remove this line a test which will follow with the patch will fail.